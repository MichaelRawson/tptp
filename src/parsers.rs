use nom::branch::alt;
use nom::bytes::complete::{escaped, tag, take_until, take_while, take_while1};
use nom::character::complete::{digit0, multispace1, one_of};
use nom::combinator::{map, opt, recognize, value};
use nom::multi::{many0, separated_list, separated_nonempty_list};
use nom::sequence::{delimited, pair, preceded, terminated, tuple};
use nom::Err::Failure;
use std::borrow::Cow;
use std::str;

use crate::syntax::*;

pub type Parsed<'a, T> = nom::IResult<&'a [u8], T, ()>;

unsafe fn to_str(bytes: &[u8]) -> &str {
    str::from_utf8_unchecked(bytes)
}

pub fn is_lower_alpha(c: u8) -> bool {
    c >= b'a' && c <= b'z'
}

pub fn is_upper_alpha(c: u8) -> bool {
    c >= b'A' && c <= b'Z'
}

pub fn is_alphanumeric(c: u8) -> bool {
    (c >= b'a' && c <= b'z')
        || (c >= b'A' && c <= b'Z')
        || (c >= b'0' && c <= b'9')
        || (c == b'_')
}

pub fn is_visible(c: u8) -> bool {
    c >= b' ' && c <= b'~'
}

pub fn is_sq_char(c: u8) -> bool {
    is_visible(c) && c != b'\'' && c != b'\\'
}

pub fn whitespace(x: &[u8]) -> Parsed<&[u8]> {
    multispace1(x)
}

pub fn comment_line(x: &[u8]) -> Parsed<&[u8]> {
    recognize(tuple((tag("%"), take_until("\n"), tag("\n"))))(x)
}

pub fn comment_block(x: &[u8]) -> Parsed<&[u8]> {
    recognize(tuple((tag("/*"), take_until("*/"), tag("*/"))))(x)
}

pub fn ignored(x: &[u8]) -> Parsed<()> {
    value((), many0(alt((whitespace, comment_line, comment_block))))(x)
}

pub fn lower_alpha(x: &[u8]) -> Parsed<&[u8]> {
    take_while1(is_lower_alpha)(x)
}

pub fn upper_alpha(x: &[u8]) -> Parsed<&[u8]> {
    take_while1(is_upper_alpha)(x)
}

pub fn alphanumeric(x: &[u8]) -> Parsed<&[u8]> {
    take_while(is_alphanumeric)(x)
}

pub fn lower_word(x: &[u8]) -> Parsed<Cow<str>> {
    map(recognize(preceded(lower_alpha, alphanumeric)), |w| {
        Cow::Borrowed(unsafe { to_str(w) })
    })(x)
}

pub fn upper_word(x: &[u8]) -> Parsed<Cow<str>> {
    map(recognize(preceded(upper_alpha, alphanumeric)), |w| {
        Cow::Borrowed(unsafe { to_str(w) })
    })(x)
}

pub fn single_quoted(x: &[u8]) -> Parsed<Cow<str>> {
    map(
        recognize(delimited(
            tag("'"),
            escaped(take_while1(is_sq_char), '\\', one_of("\\'")),
            tag("'"),
        )),
        |w| Cow::Borrowed(unsafe { to_str(w) }),
    )(x)
}

pub fn atomic_word(x: &[u8]) -> Parsed<Cow<str>> {
    alt((lower_word, single_quoted))(x)
}

pub fn integer(x: &[u8]) -> Parsed<Cow<str>> {
    map(
        recognize(preceded(
            opt(one_of("+-")),
            alt((tag("0"), preceded(one_of("123456789"), digit0))),
        )),
        |w| Cow::Borrowed(unsafe { to_str(w) }),
    )(x)
}

pub fn name(x: &[u8]) -> Parsed<Name> {
    alt((
        map(lower_word, Name::LowerWord),
        map(integer, Name::Integer),
        map(single_quoted, Name::SingleQuoted),
    ))(x)
}

pub fn name_list(x: &[u8]) -> Parsed<Vec<Name>> {
    delimited(
        pair(tag("["), ignored),
        separated_list(tuple((ignored, tag(","), ignored)), name),
        pair(ignored, tag("]")),
    )(x)
}

pub fn fof_arguments(x: &[u8]) -> Parsed<Vec<FofTerm>> {
    delimited(
        pair(tag("("), ignored),
        separated_list(tuple((ignored, tag(","), ignored)), fof_term),
        pair(ignored, tag(")")),
    )(x)
}

pub fn variable(x: &[u8]) -> Parsed<Variable> {
    map(upper_word, Variable)(x)
}

pub fn fof_plain_term(x: &[u8]) -> Parsed<FofTerm> {
    map(
        pair(name, preceded(ignored, opt(fof_arguments))),
        |(name, args)| match args {
            Some(args) => FofTerm::Functor(name, args),
            None => FofTerm::Constant(name),
        },
    )(x)
}

pub fn fof_function_term(x: &[u8]) -> Parsed<FofTerm> {
    fof_plain_term(x)
}

pub fn fof_term(x: &[u8]) -> Parsed<FofTerm> {
    alt((fof_function_term, map(variable, FofTerm::Variable)))(x)
}

pub fn fof_defined_atomic_formula(x: &[u8]) -> Parsed<FofFormula> {
    alt((
        value(FofFormula::Boolean(true), tag("$true")),
        value(FofFormula::Boolean(false), tag("$false")),
    ))(x)
}

pub fn infix_equality(x: &[u8]) -> Parsed<InfixEquality> {
    alt((
        value(InfixEquality::Equal, tag("=")),
        value(InfixEquality::NotEqual, tag("!=")),
    ))(x)
}

pub fn fof_atomic_formula(x: &[u8]) -> Parsed<FofFormula> {
    fn predicate_or_infix(x: &[u8]) -> Parsed<FofFormula> {
        let (x, (term, rest)) = pair(
            fof_term,
            opt(pair(
                preceded(ignored, infix_equality),
                preceded(ignored, fof_term),
            )),
        )(x)?;
        let result = match (term, rest) {
            (left, Some((op, right))) => Ok(FofFormula::Infix(op, left, right)),
            (FofTerm::Constant(name), None) => {
                Ok(FofFormula::Proposition(name))
            }
            (FofTerm::Functor(name, args), None) => {
                Ok(FofFormula::Predicate(name, args))
            }
            (FofTerm::Variable(_), None) => Err(Failure(())),
        }?;
        Ok((x, result))
    }

    alt((fof_defined_atomic_formula, predicate_or_infix))(x)
}

pub fn fof_quantified_formula(x: &[u8]) -> Parsed<FofFormula> {
    map(
        tuple((
            alt((
                value(FofQuantifier::Forall, tag("!")),
                value(FofQuantifier::Exists, tag("?")),
            )),
            preceded(
                tuple((ignored, tag("["), ignored)),
                separated_nonempty_list(
                    tuple((ignored, tag(","), ignored)),
                    variable,
                ),
            ),
            preceded(
                tuple((ignored, tag("]"), ignored, tag(":"), ignored)),
                fof_unit_formula,
            ),
        )),
        |(quantifier, bound, formula)| {
            FofFormula::Quantified(quantifier, bound, Box::new(formula))
        },
    )(x)
}

pub fn fof_unitary_formula(x: &[u8]) -> Parsed<FofFormula> {
    alt((
        delimited(
            pair(tag("("), ignored),
            fof_logic_formula,
            pair(ignored, tag(")")),
        ),
        fof_atomic_formula,
        fof_quantified_formula,
    ))(x)
}

pub fn fof_unary_formula(x: &[u8]) -> Parsed<FofFormula> {
    map(
        preceded(pair(tag("~"), ignored), fof_unit_formula),
        |formula| FofFormula::Unary(UnaryConnective::Not, Box::new(formula)),
    )(x)
}

pub fn fof_unit_formula(x: &[u8]) -> Parsed<FofFormula> {
    alt((fof_unary_formula, fof_unitary_formula))(x)
}

pub fn nonassoc_connective(x: &[u8]) -> Parsed<NonAssocConnective> {
    alt((
        value(NonAssocConnective::LRImplies, tag("=>")),
        value(NonAssocConnective::Equivalent, tag("<=>")),
        value(NonAssocConnective::RLImplies, tag("<=")),
        value(NonAssocConnective::NotEquivalent, tag("<~>")),
        value(NonAssocConnective::NotAnd, tag("~&")),
        value(NonAssocConnective::NotOr, tag("~|")),
    ))(x)
}

pub fn assoc_connective(x: &[u8]) -> Parsed<AssocConnective> {
    alt((
        value(AssocConnective::And, tag("&")),
        value(AssocConnective::Or, tag("|")),
    ))(x)
}

pub fn fof_logic_formula(x: &[u8]) -> Parsed<FofFormula> {
    let (x, first) =
        terminated(alt((fof_unary_formula, fof_unit_formula)), ignored)(x)?;

    if let Ok((x, op)) = nonassoc_connective(x) {
        let (x, second) = preceded(ignored, fof_unit_formula)(x)?;
        Ok((
            x,
            FofFormula::NonAssoc(op, Box::new(first), Box::new(second)),
        ))
    } else if let Ok((x, op)) = assoc_connective(x) {
        let (mut x, second) = preceded(ignored, fof_unit_formula)(x)?;
        let mut formulae = vec![first, second];
        while let Ok((after, (check_op, formula))) = pair(
            preceded(ignored, assoc_connective),
            preceded(ignored, fof_unit_formula),
        )(x)
        {
            if op == check_op {
                formulae.push(formula);
                x = after;
            } else {
                return Err(Failure(()));
            }
        }

        Ok((x, FofFormula::Assoc(op, formulae)))
    } else {
        Ok((x, first))
    }
}

pub fn fof_formula(x: &[u8]) -> Parsed<FofFormula> {
    fof_logic_formula(x)
}

pub fn literal(x: &[u8]) -> Parsed<Literal> {
    alt((
        map(
            preceded(pair(tag("~"), ignored), fof_atomic_formula),
            Literal::NegatedLiteral,
        ),
        map(fof_atomic_formula, |l| match l {
            FofFormula::Infix(_, _, _) => Literal::EqualityLiteral(l),
            _ => Literal::Literal(l),
        }),
    ))(x)
}

pub fn disjunction(x: &[u8]) -> Parsed<Vec<Literal>> {
    separated_nonempty_list(tuple((ignored, tag("|"), ignored)), literal)(x)
}

pub fn cnf_formula(x: &[u8]) -> Parsed<CnfFormula> {
    map(
        alt((
            delimited(
                pair(tag("("), ignored),
                disjunction,
                pair(ignored, tag(")")),
            ),
            disjunction,
        )),
        CnfFormula,
    )(x)
}

pub fn formula_role(x: &[u8]) -> Parsed<FormulaRole> {
    alt((
        value(FormulaRole::Axiom, tag("axiom")),
        value(FormulaRole::Hypothesis, tag("hypothesis")),
        value(FormulaRole::Definition, tag("definition")),
        value(FormulaRole::Assumption, tag("assumption")),
        value(FormulaRole::Lemma, tag("lemma")),
        value(FormulaRole::Theorem, tag("theorem")),
        value(FormulaRole::Corollary, tag("corollary")),
        value(FormulaRole::Conjecture, tag("conjecture")),
        value(FormulaRole::NegatedConjecture, tag("negated_conjecture")),
        value(FormulaRole::Plain, tag("plain")),
        value(FormulaRole::Unknown, tag("unknown")),
    ))(x)
}

pub fn unknown_source(x: &[u8]) -> Parsed<Source> {
    value(Source::Unknown, tag("unknown"))(x)
}

pub fn dag_source(x: &[u8]) -> Parsed<DagSource> {
    alt((
        map(
            pair(
                preceded(
                    tuple((tag("inference"), ignored, tag("("), ignored)),
                    atomic_word,
                ),
                delimited(
                    tuple((
                        ignored,
                        tag(","),
                        ignored,
                        tag("["),
                        ignored,
                        tag("]"),
                        ignored,
                        tag(","),
                        ignored,
                    )),
                    sources,
                    pair(ignored, tag(")")),
                ),
            ),
            |(rule, parents)| DagSource::Inference(rule, parents),
        ),
        map(name, DagSource::Name),
    ))(x)
}

pub fn external_source(x: &[u8]) -> Parsed<ExternalSource> {
    map(
        delimited(
            tuple((tag("file"), ignored, tag("("), ignored)),
            pair(
                single_quoted,
                opt(preceded(tuple((ignored, tag(","), ignored)), name)),
            ),
            pair(ignored, tag(")")),
        ),
        |(file, info)| ExternalSource::File(file, info),
    )(x)
}

pub fn sources(x: &[u8]) -> Parsed<Vec<Source>> {
    delimited(
        pair(tag("["), ignored),
        separated_list(tuple((ignored, tag(","), ignored)), source),
        pair(ignored, tag("]")),
    )(x)
}

pub fn source(x: &[u8]) -> Parsed<Source> {
    alt((
        unknown_source,
        map(external_source, Source::External),
        map(dag_source, Source::Dag),
        map(sources, Source::Sources),
    ))(x)
}

pub fn annotations(x: &[u8]) -> Parsed<Annotations> {
    map(source, |source| Annotations { source })(x)
}

pub fn include_path(x: &[u8]) -> Parsed<Included> {
    map(
        delimited(
            tag("'"),
            escaped(take_while1(is_sq_char), '\\', one_of("\\'")),
            tag("'"),
        ),
        |w| Included(unsafe { to_str(w) }.into()),
    )(x)
}

pub fn include(x: &[u8]) -> Parsed<Statement> {
    map(
        delimited(
            tuple((tag("include"), ignored, tag("("), ignored)),
            pair(
                include_path,
                opt(preceded(tuple((ignored, tag(","), ignored)), name_list)),
            ),
            pair(ignored, tag(")")),
        ),
        |(included, selection)| Statement::Include(included, selection),
    )(x)
}

pub fn fof_annotated(x: &[u8]) -> Parsed<Statement> {
    map(
        delimited(
            tuple((tag("fof"), ignored, tag("("), ignored)),
            tuple((
                name,
                preceded(tuple((ignored, tag(","), ignored)), formula_role),
                preceded(tuple((ignored, tag(","), ignored)), fof_formula),
                opt(preceded(tuple((ignored, tag(","), ignored)), annotations)),
            )),
            pair(ignored, tag(")")),
        ),
        |(name, role, formula, annotations)| {
            Statement::Fof(name, role, formula, annotations)
        },
    )(x)
}

pub fn cnf_annotated(x: &[u8]) -> Parsed<Statement> {
    map(
        delimited(
            tuple((tag("cnf"), ignored, tag("("), ignored)),
            tuple((
                name,
                preceded(tuple((ignored, tag(","), ignored)), formula_role),
                preceded(tuple((ignored, tag(","), ignored)), cnf_formula),
                opt(preceded(tuple((ignored, tag(","), ignored)), annotations)),
            )),
            pair(ignored, tag(")")),
        ),
        |(name, role, formula, annotations)| {
            Statement::Cnf(name, role, formula, annotations)
        },
    )(x)
}

pub fn tptp_input(x: &[u8]) -> Parsed<Statement> {
    terminated(
        alt((include, fof_annotated, cnf_annotated)),
        pair(ignored, tag(".")),
    )(x)
}

pub fn tptp_input_or_eof(x: &[u8]) -> Parsed<Option<Statement>> {
    if x.is_empty() {
        Ok((x, None))
    } else {
        map(delimited(ignored, tptp_input, ignored), Some)(x)
    }
}
