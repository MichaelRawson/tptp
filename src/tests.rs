use alloc::fmt;
use alloc::format;
use insta::{assert_debug_snapshot, assert_display_snapshot};

use crate::parsers::*;
use crate::syntax::*;

fn check_size<T>() {
    assert!(core::mem::size_of::<T>() <= 64);
}

fn do_parse<'a, P, T: 'a>(parser: P, input: &'a [u8]) -> T
where
    P: FnOnce(&'a [u8]) -> nom::IResult<&[u8], T, ()>,
{
    match parser(input) {
        Ok((b"\0", result)) => result,
        Ok((_, _)) => panic!("parsed, but bytes remaining"),
        Err(_) => panic!("parse error"),
    }
}

fn parse_nodisplay<'a, P, T: 'a>(parser: P, input: &'a [u8])
where
    P: FnOnce(&'a [u8]) -> nom::IResult<&[u8], T, ()>,
    T: fmt::Debug,
{
    let parsed = do_parse(parser, input);
    assert_debug_snapshot!(parsed);
}

fn parse<'a, P, T: 'a>(parser: P, input: &'a [u8])
where
    P: FnOnce(&'a [u8]) -> nom::IResult<&[u8], T, ()>,
    T: fmt::Debug + fmt::Display,
{
    let parsed = do_parse(parser, input);
    assert_debug_snapshot!(parsed);
    assert_display_snapshot!(parsed);
}

#[test]
fn test_whitespace() {
    parse_nodisplay(whitespace, b" \t\n\0");
}

#[test]
fn test_comment_line() {
    parse_nodisplay(comment_line, b"% a comment\n\0");
}

#[test]
fn test_comment_block() {
    parse_nodisplay(comment_block, b"/* a\n block * / comment\n*/\0");
}

#[test]
fn test_ignored() {
    parse_nodisplay(ignored, b"\0");
    parse_nodisplay(ignored, b"   %test\n  /* test\ntest */  \0");
}

#[test]
fn test_lower_word() {
    check_size::<LowerWord>();
    parse(lower_word, b"x\0");
    parse(lower_word, b"aA123\0");
}

#[test]
fn test_upper_word() {
    check_size::<UpperWord>();
    parse(upper_word, b"X\0");
    parse(upper_word, b"Aa123\0");
}

#[test]
fn test_dollar_word() {
    check_size::<DollarWord>();
    parse(dollar_word, b"$dollar\0");
}

#[test]
fn test_dollar_dollar_word() {
    check_size::<DollarDollarWord>();
    parse(dollar_dollar_word, b"$$dollar\0");
}

#[test]
fn test_single_quoted() {
    check_size::<SingleQuoted>();
    parse(single_quoted, b"'single quoted'\0");
    parse(single_quoted, b"'\\'\\\\'\0");
}

#[test]
fn test_distinct_object() {
    check_size::<DistinctObject>();
    parse(distinct_object, b"\"distinct object\"\0");
    parse(distinct_object, b"\"\\\"\\\\\"\0");
}

#[test]
fn test_atomic_word() {
    check_size::<AtomicWord>();
    parse(atomic_word, b"x\0");
    parse(atomic_word, b"'single quoted'\0");
}

#[test]
fn test_integer() {
    check_size::<Integer>();
    parse(integer, b"0\0");
    parse(integer, b"123\0");
    parse(integer, b"-123\0");
}

#[test]
fn test_rational() {
    check_size::<Rational>();
    parse(rational, b"0/1\0");
    parse(rational, b"123/456\0");
    parse(rational, b"-123/456\0");
}

#[test]
fn test_real() {
    check_size::<Real>();
    parse(real, b"0.0\0");
    parse(real, b"1E0\0");
    parse(real, b"-1.23E-456\0");
    parse(real, b"1e-06\0");
}

#[test]
fn test_number() {
    check_size::<Number>();
    parse(number, b"-123\0");
    parse(number, b"-123/456\0");
    parse(number, b"-1.23E-456\0");
}

#[test]
fn test_name() {
    check_size::<Name>();
    parse(name, b"lower_word2\0");
    parse(name, b"'single quoted'\0");
    parse(name, b"123\0");
}

#[test]
fn test_variable() {
    check_size::<Variable>();
    parse(variable, b"X\0");
}

#[test]
fn test_atomic_system_word() {
    check_size::<AtomicSystemWord>();
    parse(atomic_system_word, b"$$atomic\0");
}

#[test]
fn test_atomic_defined_word() {
    check_size::<AtomicDefinedWord>();
    parse(atomic_defined_word, b"$atomic\0");
}

#[test]
fn test_system_functor() {
    check_size::<SystemFunctor>();
    parse(system_functor, b"$$system_functor\0");
}

#[test]
fn test_system_constant() {
    check_size::<SystemConstant>();
    parse(system_constant, b"$$system_constant\0");
}

#[test]
fn test_defined_functor() {
    check_size::<DefinedFunctor>();
    parse(defined_functor, b"$defined_functor\0");
}

#[test]
fn test_defined_constant() {
    check_size::<DefinedConstant>();
    parse(defined_constant, b"$defined_constant\0");
}

#[test]
fn test_defined_term() {
    check_size::<DefinedTerm>();
    parse(defined_term, b"-123\0");
    parse(defined_term, b"\"distinct object\"\0");
}

#[test]
fn test_functor() {
    check_size::<Functor>();
    parse(functor, b"functor\0");
}

#[test]
fn test_constant() {
    check_size::<Constant>();
    parse(constant, b"constant\0");
}

#[test]
fn test_fof_arguments() {
    check_size::<FofArguments>();
    parse(fof_arguments, b"( c )\0");
    parse(fof_arguments, b"( X )\0");
    parse(fof_arguments, b"( X, f ( X ) )\0");
}

#[test]
fn test_fof_plain_term() {
    check_size::<FofPlainTerm>();
    parse(fof_plain_term, b"c\0");
    parse(fof_plain_term, b"f ( X )\0");
    parse(fof_plain_term, b"f ( X, g ( Y ) )\0");
}

#[test]
fn test_fof_system_term() {
    check_size::<FofSystemTerm>();
    parse(fof_system_term, b"$$c\0");
    parse(fof_system_term, b"$$f ( X )\0");
    parse(fof_system_term, b"$$f ( X, g ( Y ) )\0");
}

#[test]
fn test_fof_defined_plain_term() {
    check_size::<FofDefinedPlainTerm>();
    parse(fof_defined_plain_term, b"$defined_plain_term\0");
}

#[test]
fn test_fof_defined_atomic_term() {
    check_size::<FofDefinedAtomicTerm>();
    parse(fof_defined_atomic_term, b"$defined_atomic_term\0");
}

#[test]
fn test_fof_defined_term() {
    check_size::<FofDefinedTerm>();
    parse(fof_defined_term, b"$defined_term\0");
    parse(fof_defined_term, b"-123\0");
}

#[test]
fn test_fof_function_term() {
    check_size::<FofFunctionTerm>();
    parse(fof_function_term, b"f(X)\0");
    parse(fof_function_term, b"$defined\0");
}

#[test]
fn test_fof_term() {
    check_size::<FofTerm>();
    parse(fof_term, b"f(X)\0");
    parse(fof_term, b"X\0");
}

#[test]
fn test_fof_plain_atomic_formula() {
    check_size::<FofPlainAtomicFormula>();
    parse(fof_plain_atomic_formula, b"f ( X, g ( Y ) )\0");
}

#[test]
fn test_fof_defined_plain_formula() {
    check_size::<FofDefinedPlainFormula>();
    parse(fof_defined_plain_formula, b"$defined_plain_formula\0");
}

#[test]
fn test_infix_equality() {
    check_size::<InfixEquality>();
    parse(infix_equality, b"=\0");
}

#[test]
fn test_infix_inequality() {
    check_size::<InfixInequality>();
    parse(infix_inequality, b"!=\0");
}

#[test]
fn test_unary_connective() {
    check_size::<UnaryConnective>();
    parse(unary_connective, b"~\0");
}

#[test]
fn test_nonassoc_connective() {
    check_size::<NonassocConnective>();
    parse(nonassoc_connective, b"<=\0");
    parse(nonassoc_connective, b"<=>\0");
    parse(nonassoc_connective, b"=>\0");
    parse(nonassoc_connective, b"<~>\0");
    parse(nonassoc_connective, b"~&\0");
    parse(nonassoc_connective, b"~|\0");
}

#[test]
fn test_fof_quantifier() {
    check_size::<FofQuantifier>();
    parse(fof_quantifier, b"!\0");
    parse(fof_quantifier, b"?\0");
}

#[test]
fn test_defined_infix_pred() {
    check_size::<DefinedInfixPred>();
    parse(defined_infix_pred, b"=\0");
}

#[test]
fn test_fof_defined_infix_formula() {
    check_size::<FofDefinedInfixFormula>();
    parse(fof_defined_infix_formula, b"f(X) = c\0");
}

#[test]
fn test_fof_defined_atomic_formula() {
    check_size::<FofDefinedAtomicFormula>();
    parse(fof_defined_atomic_formula, b"$true\0");
    parse(fof_defined_atomic_formula, b"$false\0");
    parse(fof_defined_atomic_formula, b"f(X) = c\0");
}

#[test]
fn test_fof_system_atomic_formula() {
    check_size::<FofSystemAtomicFormula>();
    parse(fof_system_atomic_formula, b"$$system\0");
}

#[test]
fn test_fof_atomic_formula() {
    check_size::<FofAtomicFormula>();
    parse(fof_atomic_formula, b"$true\0");
    parse(fof_atomic_formula, b"f(X) = Y\0");
    parse(fof_atomic_formula, b"p(X)\0");
    parse(fof_atomic_formula, b"$$system\0");
}

#[test]
fn test_fof_variable_list() {
    check_size::<FofVariableList>();
    parse(fof_variable_list, b"X\0");
    parse(fof_variable_list, b"X , Y\0");
    parse(fof_variable_list, b"X , Y , Z\0");
}

#[test]
fn test_fof_quantified_formula() {
    check_size::<FofQuantifiedFormula>();
    parse(fof_quantified_formula, b"! [ X ] : $true\0");
    parse(fof_quantified_formula, b"? [ X , Y , Z ] : $true\0");
}

#[test]
fn test_fof_infix_unary() {
    check_size::<FofInfixUnary>();
    parse(fof_infix_unary, b"f(X) != c\0");
}

#[test]
fn test_fof_unary_formula() {
    check_size::<FofUnaryFormula>();
    parse(fof_unary_formula, b"~ $true\0");
    parse(fof_unary_formula, b"f(X) != c\0");
}

#[test]
fn test_fof_unitary_formula() {
    check_size::<FofUnitaryFormula>();
    parse(fof_unitary_formula, b"( $true )\0");
    parse(fof_unitary_formula, b"$true\0");
    parse(fof_unitary_formula, b"![X]: $true\0");
}

#[test]
fn test_fof_unit_formula() {
    check_size::<FofUnitFormula>();
    parse(fof_unit_formula, b"($true)\0");
    parse(fof_unit_formula, b"~$true\0");
}

#[test]
fn test_fof_binary_nonassoc() {
    check_size::<FofBinaryNonassoc>();
    parse(fof_binary_nonassoc, b"p => q\0");
    parse(fof_binary_nonassoc, b"p ~| q\0");
}

#[test]
fn test_fof_or_formula() {
    check_size::<FofOrFormula>();
    parse(fof_or_formula, b"p | q | r\0");
}

#[test]
fn test_fof_and_formula() {
    check_size::<FofAndFormula>();
    parse(fof_and_formula, b"p & q\0");
}

#[test]
fn test_fof_binary_assoc() {
    check_size::<FofBinaryAssoc>();
    parse(fof_binary_assoc, b"p | q | r\0");
    parse(fof_binary_assoc, b"p & q\0");
}

#[test]
fn test_fof_logic_formula() {
    check_size::<FofLogicFormula>();
    parse(fof_logic_formula, b"~p\0");
    parse(fof_logic_formula, b"p => q\0");
    parse(fof_logic_formula, b"p & q\0");
    parse(fof_logic_formula, b"p | q | r\0");
    parse(fof_logic_formula, b"p\0");
    parse(fof_logic_formula, b"~p => q\0");
}

#[test]
fn test_fof_formula() {
    check_size::<FofFormula>();
    parse(fof_formula, b"p\0");
    parse(fof_formula, b"~~p\0");
    parse(fof_formula, b"(p)\0");
    parse(fof_formula, b"$true|$false\0");
    parse(fof_formula, b"(![X,Y,Z]:?[Q]:Q!=p(A))&p&(q=>r)\0");
}

#[test]
fn test_literal() {
    check_size::<Literal>();
    parse(literal, b"p\0");
    parse(literal, b"~ p\0");
    parse(literal, b"f(X) = c\0");
}

#[test]
fn test_disjunction() {
    check_size::<Disjunction>();
    parse(disjunction, b"p\0");
    parse(disjunction, b"p | ~q\0");
    parse(disjunction, b"p | ~q | r\0");
}

#[test]
fn test_cnf_formula() {
    check_size::<CnfFormula>();
    parse(cnf_formula, b"p\0");
    parse(cnf_formula, b"( p )\0");
    parse(cnf_formula, b"p | ~q\0");
}

#[test]
fn test_role() {
    check_size::<FormulaRole>();
    parse(formula_role, b"axiom\0");
    parse(formula_role, b"conjecture\0");
}

#[test]
fn test_general_terms() {
    check_size::<GeneralTerms>();
    parse(general_terms, b"X\0");
    parse(general_terms, b"X , Y\0");
}

#[test]
fn test_general_list() {
    check_size::<GeneralList>();
    parse(general_list, b"[ ]\0");
    parse(general_list, b"[ X ]\0");
    parse(general_list, b"[ X , Y ]\0");
}

#[test]
fn test_general_function() {
    check_size::<GeneralFunction>();
    parse(general_function, b"atomic ( X )\0");
}

#[test]
fn test_formula_data() {
    check_size::<FormulaData>();
    parse(formula_data, b"$fof ( p )\0");
    parse(formula_data, b"$cnf ( p )\0");
    parse(formula_data, b"$fot ( t )\0");
}

#[test]
fn test_general_data() {
    check_size::<GeneralData>();
    parse(general_data, b"c\0");
    parse(general_data, b"X\0");
    parse(general_data, b"atomic ( X )\0");
    parse(general_data, b"$fof ( p )\0");
    parse(general_data, b"123\0");
    parse(general_data, b"\"distinct object\"\0");
}

#[test]
fn test_general_term() {
    check_size::<GeneralTerm>();
    parse(general_term, b"[ X , Y ]\0");
    parse(general_term, b"$fof ( p )\0");
    parse(general_term, b"X : Y\0");
}

#[test]
fn test_useful_info() {
    check_size::<UsefulInfo>();
    parse(useful_info, b"[ X , Y ]\0");
}

#[test]
fn test_optional_info() {
    check_size::<OptionalInfo>();
    parse(optional_info, b"\0");
    parse(optional_info, b", [ X , Y ]\0");
}

#[test]
fn test_annotations() {
    check_size::<Annotations>();
    parse(annotations, b"\0");
    parse(annotations, b", c\0");
    parse(annotations, b", c , [X]\0");
}

#[test]
fn test_fof_annotated() {
    check_size::<FofAnnotated>();
    parse(fof_annotated, b"fof ( test , axiom , $true ) .\0");
    parse(fof_annotated, b"fof ( test , axiom , $true , unknown ) .\0");
    parse(
        fof_annotated,
        b"fof ( test , axiom , $true , unknown , [] ) .\0",
    );
}

#[test]
fn test_cnf_annotated() {
    check_size::<CnfAnnotated>();
    parse(cnf_annotated, b"cnf ( test , axiom , $true ) .\0");
    parse(cnf_annotated, b"cnf ( test , axiom , $true , unknown ) .\0");
    parse(
        cnf_annotated,
        b"cnf ( test , axiom , $true , unknown , [] ) .\0",
    );
}

#[test]
fn test_annotated_formula() {
    check_size::<AnnotatedFormula>();
    parse(annotated_formula, b"fof ( test , axiom , $true ) .\0");
    parse(annotated_formula, b"cnf ( test , axiom , $true ) .\0");
}

#[test]
fn test_file_name() {
    check_size::<FileName>();
    parse(file_name, b"'test'\0");
}

#[test]
fn test_name_list() {
    check_size::<NameList>();
    parse(name_list, b"name , 'name' , 123\0");
    parse(name_list, b"name\0");
}

#[test]
fn test_formula_selection() {
    check_size::<FormulaSelection>();
    parse(formula_selection, b", [ name , 'name' , 123 ]\0");
    parse(formula_selection, b", [ name ]\0");
}

#[test]
fn test_include() {
    check_size::<Include>();
    parse(include, b"include ( 'test' ) .\0");
    parse(include, b"include( 'test', [ test ] ) .\0");
}

#[test]
fn test_tptp_input() {
    check_size::<TPTPInput>();
    parse(tptp_input, b"include ( 'test' ) .\0");
    parse(tptp_input, b"fof ( test , axiom , $true ) .\0");
    parse(tptp_input, b"cnf ( test , axiom , $true ) .\0");
}
