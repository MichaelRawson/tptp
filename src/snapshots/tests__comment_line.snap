---
source: src/tests.rs
expression: parsed
---
[
    37,
    32,
    97,
    32,
    99,
    111,
    109,
    109,
    101,
    110,
    116,
    10,
]
